types/react/index.d.ts

```diff
@@ -560,6 +560,165 @@ declare namespace React {
 
     function forwardRef<T, P = {}>(Component: RefForwardingComponent<T, P>): ComponentType<P & ClassAttributes<T>>;
 
+    //
+    // React Hooks
+    // ----------------------------------------------------------------------
+
+    // based on the code in https://github.com/facebook/react/pull/13968
+
+    // Unlike the class component setState, the updates are not allowed to be partial
+    type SetStateAction<S> = S | ((prevState: S) => S);
+    // this technically does accept a second argument, but it's already under a deprecation warning
+    // and it's not even released so probably better to not define it.
+    type Dispatch<A> = (value: A) => void;
+    // Unlike redux, the actions _can_ be anything
+    type Reducer<S, A> = (prevState: S, action: A) => S;
+    // The identity check is done with the SameValue algorithm (Object.is), which is stricter than ===
+    // TODO (TypeScript 3.0): ReadonlyArray<unknown>
+    type InputIdentityList = ReadonlyArray<any>;
+
+    // NOTE: Currently, in alpha.0, the effect callbacks are actually allowed to return anything,
+    // but functions are treated specially. The next version published with hooks will warn if you actually
+    // return anything besides `void` or a callback. Async effects need to call an async function inside
+    // them.
+    type EffectCallback = () => (void | (() => void));
```

Comment:
Should a `never` type be used here? Because [functions returning non-void are assignable to a function returning void](https://github.com/Microsoft/TypeScript/wiki/FAQ#why-are-functions-returning-non-void-assignable-to-function-returning-void).
```suggestion
    type EffectCallback = () => (never | (() => void));
```